base/strings/basic.jl

```diff
@@ -613,6 +613,26 @@ isascii(c::Char) = bswap(reinterpret(UInt32, c)) < 0x80
 isascii(s::AbstractString) = all(isascii, s)
 isascii(c::AbstractChar) = UInt32(c) < 0x80
 
+function _isascii(code_units::AbstractVector{CU}, first, last) where {CU}
+    r = zero(CU)
+    for n = first:last
+        @inbounds r |= code_units[n]
+    end
+    return r < 0x80
+end
+
+function isascii(code_units::AbstractVector{CU}) where {CU <: Integer}
```

Comment:
Should one say the following?
```suggestion
function isascii(code_units::AbstractVector{CU}) where {CU <: Unsigned}
```
Otherwise the test `r < 0x80` in `_isascii` does not test for 7-bit integers.

Reply:
```julia
julia> Int(0x7f) < 0x80
true
```

Reply:
Perhaps, testing `0 ≤ r < 0x80` would be better. The compiler knows that testing if an unsigned integer is greater than zero is a no-op and can eliminate that work. The presumption of this code is that argument is some kind of vector of code points encoded as integers. It seems sensible to use unsigned integer types for that since code points can only be non-negative, but what if I wanted to just use `Int` for this and, e.g. let `-1` be an EOF character? Now, if you let that `-1` into this function then you'll get the result that the string isn't ASCII because of that EOF indicator, but if you don't want that to happen you'd need to make sure you don't include the EOF. The way I think about it is that logically code points are integers and any way of encoding an integer is fine; the ASCII code points are the ones from 0 to 0x7f, so test if they're in that range or not.

Reply:
> The compiler knows that testing if an unsigned integer is greater than zero is a no-op

Verified on v1.9.0-beta2. However, there seems to be a problem on master:
```
julia> f(x) = zero(x) <= x
julia> @code_native f(UInt(1))
	.text
	.file	"f"
	.globl	julia_f_204                     # -- Begin function julia_f_204
	.p2align	4, 0x90
	.type	julia_f_204,@function
julia_f_204:                            # @julia_f_204
; ┌ @ REPL[29]:1 within `f`
	.cfi_startproc
# %bb.0:                                # %top
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset rbp, -16
	mov	rbp, rsp
	.cfi_def_cfa_register rbp
	#APP
	mov	rax, qword ptr fs:[0]
	#NO_APP
	mov	rax, qword ptr [rax - 8]
	mov	rax, qword ptr [rax + 16]
	mov	rax, qword ptr [rax + 16]
	#MEMBARRIER
	mov	rax, qword ptr [rax]
	#MEMBARRIER
	mov	al, 1
	pop	rbp
	.cfi_def_cfa rsp, 8
	ret
.Lfunc_end0:
	.size	julia_f_204, .Lfunc_end0-julia_f_204
	.cfi_endproc
; └
                                        # -- End function
	.section	".note.GNU-stack","",@progbits
```
```
Julia Version 1.10.0-DEV.599
Commit dd832c0a2b (2023-02-13 22:08 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 4 × Intel(R) Core(TM) i3-10110U CPU @ 2.10GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake)
  Threads: 1 on 4 virtual cores
```

Reply:
Issue filed for the code gen regression: https://github.com/JuliaLang/julia/issues/48680.

Reply:
Most of that is compiler artifacts (as noted by Keno), it still reduces fine - see the `mov al, 1`. EVERYTHING above the `#MEMBARRIER` is noise due to how this was generated. Muting that is hard :pensive:

Reply:
I ended up using `0 ≤ r < 0x80 ` as it had no impact on the results